File: Python/lab1/src/main.py
import threading
import os
import time
from typing import List


def parallel_sum(data: List[int], n_slices: int) -> int:
    if not data or n_slices <= 0:
        return 0

    length = len(data)
    n = min(n_slices, length)
    base = length // n
    remainder = length % n

    threads = []
    partial_sums = [0] * n

    def worker(idx: int, start: int, end: int):
        partial_sum = 0
        for i in range(start, end):
            partial_sum += data[i]
        partial_sums[idx] = partial_sum

    start = 0
    for i in range(n):
        add = 1 if i < remainder else 0
        chunk_len = base + add
        end = start + chunk_len

        t = threading.Thread(target=worker, args=(i, start, end), daemon=False)
        threads.append(t)
        t.start()

        start = end

    for t in threads:
        t.join()

    total_sum = 0
    for value in partial_sums:
        total_sum += value

    return total_sum


if __name__ == "__main__":
    n_items = 2_000_000
    data = [(i % 1000) for i in range(n_items)]

    n_threads = os.cpu_count() or 4
    n_slices = n_threads
    print("Using ", n_slices, " threads / slices")

    start_parallel = time.perf_counter()
    total = parallel_sum(data, n_slices)
    end_parallel = time.perf_counter()
    duration_parallel = (end_parallel - start_parallel) * 1000
    print("Total sum is ", total)
    print(f"Elapsed time is {duration_parallel:.0f} ms")

    start_single = time.perf_counter()
    total_check = sum(data)
    end_single = time.perf_counter()
    duration_single = (end_single - start_single) * 1000
    print("Total sum check is ", total_check)
    print(f"Elapsed time is {duration_single:.0f} ms")

    assert total == total_check

File: Rust/lab1/src/main.rs
use std::sync::Arc;
use std::thread;
use std::time::Instant;

// data is wrapped in Arc, so threads can share it without making copies
// return i128 - safe bet in case of large sums
fn parallel_sum(data: Arc<Vec<i64>>, n_slices: usize) -> i128 {
    if data.is_empty() || n_slices == 0 {
        return 0;
    }

    let len = data.len();
    // do not make more slices than elements
    let n = n_slices.min(len);
    let base = len / n;  // minimal chunk size
    let remainder = len % n;  // firts remainder chunks get one more element

    let mut handlers = Vec::with_capacity(n);
    let mut start = 0usize;

    for i in 0..n {
        // distribute the remainder among first chunks
        let add = if i < remainder { 1 } else { 0 };
        let chunk_len = base + add;
        let end = start + chunk_len;

        let data_clone = data.clone();  // cheap arc pointer clone

        let handler = thread::spawn(move || -> i128 {
            let slice = &data_clone[start..end];

            let mut partial_sum = 0i128;
            for &v in slice.iter() {
                partial_sum += v as i128;
            }

            partial_sum
        });

        handlers.push(handler);
        start = end;
    }

    let mut total_sum = 0i128;
    for h in handlers {
        total_sum += h.join().expect("thread panicked");
    }

    total_sum
}

fn main() {
    let n_items = 2_000_000usize;
    let mut data_raw = Vec::with_capacity(n_items);
    for i in 0..n_items {
        data_raw.push((i % 1000) as i64);
    }

    let data = Arc::new(data_raw);

    // calculate optimal number of threads per system to use in program;
    // extract value from Result or default to 4
    let n_threads = thread::available_parallelism()
        .map(|n| n.get())
        .unwrap_or(4);

    let n_slices = n_threads;
    println!("Using {n_slices} threads / slices");

    let start_parallel = Instant::now();
    let total_sum = parallel_sum(Arc::clone(&data), n_slices);
    let duration_parallel = start_parallel.elapsed().as_millis();
    println!("Total sum is {total_sum}");
    println!("Elapsed time is {duration_parallel} ms");

    let start_single = Instant::now();
    // dereference each value from iter() and apply i128 conversion
    let total_sum_check: i128 = data.iter().map(|&x| x as i128).sum();
    let duration_single = start_single.elapsed().as_millis();
    println!("Total sum check is {total_sum_check}");
    println!("Elapsed time is {duration_single} ms");

    assert_eq!(total_sum, total_sum_check);
}

